librz/util/buf.c

```diff
@@ -418,10 +534,112 @@ RZ_API RzBuffer *rz_buf_new_with_pointers(const ut8 *bytes, ut64 len, bool steal
 	return new_buffer(RZ_BUFFER_BYTES, &u);
 }
 
-RZ_API RzBuffer *rz_buf_new_with_string(RZ_NONNULL const char *msg) {
+/**
+ * \brief Creates a new buffer from a string.
+ * \param msg The source string used to create the buffer.
+ * \return Return the new allocated buffer.
+ *
+ * The function creates a new allocated buffer using the RZ_BUFFER_BYTES back end.
+ */
+RZ_API RZ_OWN RzBuffer *rz_buf_new_with_string(RZ_NONNULL const char *msg) {
 	return rz_buf_new_with_bytes((const ut8 *)msg, (ut64)strlen(msg));
 }
 
+/**
+ * \brief Get a string whith a max length from the buffer
+ * \param b RzBuffer pointer
+ * \param addr The address of the string
+ * \param size The max length authorized
+ * \return A string with a length <= size or NULL
+ *
+ * Return an heap-allocated string read from the RzBuffer b at address addr. The
+ * length depends on the first '\0' found and the arguments size in the buffer.
+ * If there is no '\0' in the buffer, there is no string, thus NULL is returned.
+ */
+RZ_API RZ_OWN char *rz_buf_get_nstring(RzBuffer *b, ut64 addr, size_t size) {
+	rz_return_val_if_fail(b, NULL);
+
+	RzStrBuf *buf = rz_strbuf_new(NULL);
+
+	while (true) {
+		char tmp[GET_STRING_BUFFER_SIZE];
+		st64 r = rz_buf_read_at(b, addr, (ut8 *)tmp, sizeof(tmp));
+		if (r < 1) {
+			rz_strbuf_free(buf);
+			return NULL;
+		}
+
+		size_t count = strnlen(tmp, r);
+		rz_strbuf_append_n(buf, tmp, count);
+
+		if (count > size) {
+			rz_strbuf_free(buf);
+			return NULL;
+		}
+
+		if (count != r) {
+			break;
+		}
+
+		addr += r;
+		size -= count;
+	}
+
+	char *result = rz_strbuf_drain(buf);
+	return result;
+}
+
+/**
+ * \brief Get a string from the buffer
+ * \param b RzBuffer pointer
+ * \param addr The address of the string
+ * \return A string with a length <= size or NULL
+ *
+ * Return an heap-allocated string read from the RzBuffer b at address addr. The
+ * length depends on the first '\0' found in the buffer. If there is no '\0' in
+ * the buffer, there is no string, thus NULL is returned.
+ */
+RZ_API RZ_OWN char *rz_buf_get_string(RZ_NONNULL RzBuffer *b, ut64 addr) {
+	rz_return_val_if_fail(b, NULL);
+
+	return rz_buf_get_nstring(b, addr, rz_buf_size(b));
+}
+
+/**
+ * \brief Stringify the buffer
+ * \param b RzBuffer pointer
+ * \return Return an allocated string.
+ *
+ * The string is guaranteed to be NULL terminated even if the buffer doesn't have
+ * any NULL bytes.
+ */
+RZ_API RZ_OWN char *rz_buf_to_string(RZ_NONNULL RzBuffer *b) {
+	rz_return_val_if_fail(b, NULL);
+
+	ut64 size = rz_buf_size(b);
+	char *result = RZ_NEWS(char, size + 1);
+	if (!result) {
+		return NULL;
+	}
+
+	if (rz_buf_read_at(b, 0, (ut8 *)result, size) < 0) {
+		free(result);
+		return NULL;
+	}
+
+	result[size] = '\0';
+
+	return result;
+}
+
+/**
+ * \brief Return a reference to the source buffer.
+ * \param b ...
+ * \return Return the new referenced buffer.
+ *
+ * The function increment the reference counting and return a new instance of the
+ * object.
```

Comment:
* \brief Increment the reference count of the buffer
* \param b RzBuffer to reference
 * \return Return the same value \p b
 *
 * The function increment the reference count of the buffer